src/macaron/repo_finder/provenance_finder.py

```diff
@@ -35,191 +37,350 @@ def __init__(self) -> None:
-                download_path = os.path.join(temp_dir, f"{artifact_id}.intoto.jsonl")
-                if not npm_provenance_asset.download(download_path):
-                    logger.debug("Unable to find an npm provenance for %s@%s", artifact_id, version)
-                    return None
-
-                try:
-                    # Load the provenance file.
-                    provenance_payload = load_provenance_payload(download_path)
-                except LoadIntotoAttestationError as loadintotoerror:
-                    logger.error("Error while loading provenance %s", loadintotoerror)
-                    return None
-
-                return provenance_payload
-        except OSError as error:
-            logger.error("Error while storing provenance in the temporary directory: %s", error)
-            return None
-
-    @staticmethod
-    def find_gav_provenance(purl: PackageURL, jfrog_registry: JFrogMavenRegistry) -> InTotoPayload | None:
-        """Find and download the GAV based provenance for the passed PURL.
-
-        Parameters
-        ----------
-        purl: PackageURL
-            The PURL of the analysis target.
-        jfrog_registry: JFrogMavenRegistry
-            The JFrog registry to find provenance in.
-
-        Returns
-        -------
-        InTotoPayload | None
-            The provenance payload if found, or None.
+        if not verification_function:
+            if determine_abstract_purl_type(purl) == AbstractPurlType.REPOSITORY:
+                # Do not perform default verification for repository type targets.
+                return False
+
+            if purl.type == "npm":
+                verification_function = verify_npm_provenance
+                parameters = [purl, provenance]
+            else:
+                logger.debug("Provenance verification not supported for PURL type: %s", purl.type)
+
+        if not verification_function:
+            logger.debug("No provenance verification function provided/found for %s", purl)
+            return False
+
+        if not parameters:
+            logger.debug("No parameter arguments for provenance verification function.")
+
+        if parameters:
+            provenance_verified = verification_function(*parameters)
+        else:
+            # This will never be reached, and exists only for mypy.
+            return False
+        if not provenance_verified:
+            logger.debug("Provenance could not be verified.")
+        return provenance_verified
+
+
+def find_npm_provenance(purl: PackageURL, registry: NPMRegistry) -> list[InTotoPayload]:
+    """Find and download the NPM based provenance for the passed PURL.
+
+    Two kinds of attestation can be retrieved from npm: "Provenance" and "Publish". The "Provenance" attestation
+     contains the important information Macaron seeks, while the "Publish" attestation is pre-verified. Comparison
+     of the signed vs unsigned at the subject level, allows the unsigned to be verified too.
```

Comment:
nit: I think it would be clearer if we mention which of `Provenance` and `Publish` is signed, and which of them is unsigned.